packages/libs/chainweb-stream-client/src/sliding-cache.ts

```diff
@@ -0,0 +1,176 @@
+import { ITransaction } from './types';
+
+/*
+ *  A De-duplicating cache that keeps the most recent N (.cacheSpan) elements by some attribute value (.cacheGetter callback return value)
+ *    - for cw-stream-client:
+ *      cacheSpan is the window over which we may receive duplicates when reconnecting, e.g. CONFIRMATION_DEPTH + 3
+ *      cacheGetter returns txn => txn.height
+ *  - It must support updating elements
+ *    - for cw-stream-client: same txn can be received multiple times with different confirmation depths
+ *  - An element's identity is returned by identityCheck
+ *    - for cw-stream-client: txn => txn.meta.id
+ *  - Since elements can be updated, equality between elements with the same identity is returned by equalityCheck callback
+ *    - for cw-stream-client: txn => txn.meta.id && txn.meta.confirmations
+ *
+ *  - call .addCache([txn1, txn2, txn3])
+ *    returns bool[] like [shouldUpdate1, shouldUpdate2, shouldUpdate3]
+ *      each bool array idx corresponds to the input array element of the same idx
+ *      true iff a consumer event should be triggered, i.e. it is all of:
+ *        - within the span we are tracking, e.g. height is within [max_height - 12, max_height]
+ *        - has not seen before in its exact identity+equality configuration
+ *
+ *  Note: identity vs equality is used internally to distinguish between "push new element to cache" (no identity match found)
+ *  vs "find this item in cache and update it to the latest value (identity match found, did not equal latest version exactly)
+ *
+ */
+
+type SlidingCacheConfig<Shape extends ITransaction> = Pick<
+  SlidingCache<Shape>,
+  'cacheGetter' | 'cacheSpan' | 'identityCheck' | 'equalityCheck'
+>;
+
+type SlidingCacheConstructor<Shape extends ITransaction> = Partial<
+  SlidingCacheConfig<Shape>
+>;
+
+const defaults: SlidingCacheConfig<ITransaction> = {
+  cacheGetter: ({ height }) => height,
+  cacheSpan: 3,
+  identityCheck: (a, b) => a.meta?.id === b.meta?.id,
+  equalityCheck: (a, b) =>
+    a.meta?.id === b.meta?.id &&
+    a.meta?.confirmations === b.meta?.confirmations,
+};
+
+export default class SlidingCache<Shape extends ITransaction>
+  implements SlidingCacheConfig<Shape>
+{
+  public minCacheValue: number = Infinity;
+  public maxCacheValue: number = -Infinity;
+  public cache: Shape[] = [];
```

Comment:
Depending on the `cacheSpan` this can get a very expensive. It could be more efficient to use the identifier as a key in a Map (opposed to a obj, as you're more `getting` from the Map/obj to compare, than `setting` to the map/obj)

Reply:
I do agree with this but I think there is a tradeoff to be made for the first release

For now I would opt for a correct implementation without over-complicating the data structures

In theory we don't need to store the entire txns either, we could keep 3 things per element: identity value (txn id), equality value (txn id + confirmation depth) and the caching value (height)

So maybe something like 2 maps, one with `{ [height]: [txn_id1, txn_id2, txn_id3] }` that would be necessary to evict quickly, and another with `{ [txn_id1]: [txn1_confirmation_depth] }` to track the most recent version of each transaction 

But that may also have some overhead as we will be creating new objects, whereas with this pattern we should just be referencing the existing txn in memory. 

I will test how this works with the chattiest module I can think (`coin`) and see if it releasable in this state or not.